### ISSUE TITLE
Unexpected type error when passing generic function to higher order function

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#30691]
**TypeScript Version:**  3.4.1

**Search Terms:** higher order function generic type inference

**Code**

```ts
declare const withDefaults: <Params, DefaultParamsKeys extends keyof Params>(
    fn: (params: Params) => void,
    defaultParams: Pick<Params, DefaultParamsKeys>,
) => void;

declare const fn1: (options: { required: number; optional?: number }) => void;

// No type error, good!
withDefaults(fn1, {
    optional: 1,
});

// Same as above, but generic
declare const fn2: <P>(options: { required: number; optional?: number }) => void;

// Unexpected type error:
/* Argument of type '<P>(options: { required: number; optional?: number | undefined; }) => void' is not assignable to parameter of type '(params: { optional: number; }) => void'.
  Types of parameters 'options' and 'params' are incompatible.
    Property 'required' is missing in type '{ optional: number; }' but required in type '{ required: number; optional?: number | undefined; }'. */
withDefaults(fn2, {
    optional: 1,
});
```

**Comment by RyanCavanaugh** [comment]
I believe the root cause here is that a generic function signature is a lower-priority inference site than a real argument, which definitely makes sense for signatures that *aren't* needlessly generic.

Some real use case would be preferable here in terms of recommended next steps.

**Comment by OliverJAsh** [comment]
> Some real use case would be preferable here in terms of recommended next steps.

To elaborate on my above example, we have a `withDefaults` higher order function which allows us to apply default parameters to a function. A copy of the provided function will be returned with the default parameters applied.

This works fine with non-generic functions, but as per this issue, we've ran into problems when using generic functions. Full example:

``` ts
type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
type Diff<T, K extends keyof T> = Omit<T, K> & Partial<T>;

declare const withDefaults: <
  Params extends object,
  DefaultParamsKeys extends keyof Params,
  Result
>(
  fn: (params: Params) => Result,
  defaultParams: Pick<Params, DefaultParamsKeys>,
) => (inputParams: Diff<Params, keyof typeof defaultParams>) => Result

declare const fn: <P>(options: { required: number; optional?: number; props: P }) => void;

// Unexpected type error:
/* Argument of type '<P>(options: { required: number; optional?: number | undefined; props: P; }) => void' is not assignable to parameter of type '(params: { optional: number; }) => void'.
  Types of parameters 'options' and 'params' are incompatible.
    Type '{ optional: number; }' is missing the following properties from type '{ required: number; optional?: number | undefined; props: {}; }': required, props */
withDefaults(fn, {
    optional: 1,
});
```

You might ask why we use this instead of specifying defaults when destructuring. There are good reasons but I'll have to collect my thoughts—let me know if you do need that information and I'll see what I can put together.

**Comment by RyanCavanaugh** [comment]
I'm going to assume you've figured out something better in the meantime 😶
